bigquery_etl/dryrun.py

```diff
@@ -221,42 +244,124 @@ def get_referenced_tables(self):
+
+            if "date" in date_filter:
+                filtered_content = (
+                    f"{self.get_sql()}WHERE {date_filter} > current_date()"
+                )
+                if (
+                    DryRun(self.sqlfile, filtered_content).get_error()
+                    == Errors.DATE_FILTER_NEEDED_AND_SYNTAX
+                ):
+                    # If the date filter (e.g. WHERE crash_date > current_date())
+                    # is added to a query that already has a WHERE clause,
+                    # it will throw an error. To fix this, we need to
+                    # append 'AND' instead of 'WHERE'
+                    filtered_content = (
+                        f"{self.get_sql()}AND {date_filter} > current_date()"
+                    )
+
+            if "timestamp" in date_filter:
+                filtered_content = (
+                    f"{self.get_sql()}WHERE {date_filter} > current_timestamp()"
+                )
+                if (
+                    DryRun(sqlfile=self.sqlfile, content=filtered_content).get_error()
+                    == Errors.DATE_FILTER_NEEDED_AND_SYNTAX
+                ):
+                    filtered_content = (
+                        f"{self.get_sql()}AND {date_filter} > current_timestamp()"
+                    )
+
+            stripped_dml_result = DryRun(sqlfile=self.sqlfile, content=filtered_content)
+            if (
+                stripped_dml_result.get_error() == None
+                and "referencedTables" in stripped_dml_result.dry_run_result
+            ):
+                return stripped_dml_result.dry_run_result["referencedTables"]
+
         return []
 
     def is_valid(self):
         """Dry run the provided SQL file and check if valid."""
         if self.dry_run_result is None:
             return False
 
-        if "errors" in self.dry_run_result and len(self.dry_run_result["errors"]) == 1:
-            error = self.dry_run_result["errors"][0]
-        else:
-            error = None
-
         if self.dry_run_result["valid"]:
             print(f"{self.sqlfile:59} OK")
-        elif (
-            error
-            and error.get("code", None) in [400, 403]
-            and "does not have bigquery.tables.create permission for dataset"
-            in error.get("message", "")
-        ):
+        elif self.get_error() == Errors.READ_ONLY:
             # We want the dryrun service to only have read permissions, so
             # we expect CREATE VIEW and CREATE TABLE to throw specific
             # exceptions.
             print(f"{self.sqlfile:59} OK")
+        elif self.get_error() == Errors.DATE_FILTER_NEEDED and self.strip_dml:
+            # Some queries require a partition filter
+            # (submission_date, submission_timestamp, etc.)
+            # We mark these requests as valid and add a date filter
+            # in get_referenced_table()
+            print(f"{self.sqlfile:59} OK but DATE FILTER NEEDED")
```

Comment:
This is changing the behavior of the dry run because jobs that would have previously failed now pass. I don't think this should cause any issues though.

Reply:
I believe it would only change the behavior of jobs that have `strip_dml` flag on, so it shouldn't affect any existing jobs